### ISSUE TITLE
False message about lack of unsafe block

### ISSUE THREAD

**Issue by Tearth** [rust-analyzer/rust-analyzer#10665]
I've spotted a problem with a false error when using UnsafeCell and Vector - in the code below, rust-analyzer tells that `for entry in unsafe { (*unsafe_cell.get())[index].entries } {` lacks unsafe block, even if it's present and it compiles/runs without any issue. Is it intended behavior?

rust-analyzer version: v0.2.792

```rust
use std::cell::UnsafeCell;

#[derive(Clone, Copy)]
struct PerftHashTableBucket {
    pub entries: [PerftHashTableEntry; 4],
}

#[derive(Clone, Copy)]
pub struct PerftHashTableEntry {
    pub key_and_depth: u64,
    pub leafs_count: u64,
}

fn main() {
    let unsafe_cell: UnsafeCell<Vec<PerftHashTableBucket>> = UnsafeCell::new(Vec::with_capacity(100));

    for index in 0..4 {
        for entry in unsafe { (*unsafe_cell.get())[index].entries } {
            println!("{} {}", entry.key_and_depth, entry.leafs_count);
        }
    }
}
```

![image](https://user-images.githubusercontent.com/26147774/139536122-32e005f6-2980-41d5-9be7-0963617e3aac.png)

**Comment by Veykril** [comment]
Ah this is running into the for loops mistaking the iterable expression as the loop body due to the BlockExpr usage 😬

**Comment by lnicola** [comment]
Minimized:

```rust
unsafe fn f() -> Vec<i32> {
    vec![]
}

fn main() {
    for x in unsafe { f() } {}
}
```
